### ISSUE TITLE
Singleton components

### ISSUE THREAD

**Issue by cloneable** [bevyengine/bevy#8944]
## What problem does this solve or what need does it fill?

I'd like to ask for support of singleton/unique components, i.e. components with at most one instance, that can be used for quickly changing markers of entities (like `PrimaryWindow`), optionally carrying data.

## What solution would you like?

A new `ComponentStorage` implementation that stores at most one instance, though API-wise I suggest using distinct parameters in the `#[component]` attribute macro, like `#[component(unique = true)]` instead of `#[component(storage = "UniqueSet")]` because a storage type by itself should not mandate cardinality.
Adding a unique component to an entity will detach it from another entity if set, allowing fast moving markers in one step.

## What alternative(s) have you considered?

* There's some overlap with `Resource`s/`Local`s storing an `Entity` reference, but singleton components integrate better with queries and query filters.
* One can use `Query::get_single` on a `SparseSet`, but this doesn't guarantee true uniqueness of a component.

**Comment by nicopap** [comment]
The important behavior is a bit buried in your description, so let me quote it.

> Adding a unique component to an entity will detach it from another entity if set

**Comment by CptPotato** [comment]
Couldn't this be a resource that references the entity it's "attached" to?

**Comment by cloneable** [comment]
> Couldn't this be a resource that references the entity it's "attached" to?

Yes, totally, I listed this under alternatives. Design-wise it's better to use a component for something that's related to entities, and practically it will simplify systems and let Bevy ECS do the work of mapping an entity to a singleton component and vice versa.
